### ISSUE TITLE
TaskEditDialog seems to reset the "user account" setting

### ISSUE THREAD

**Issue by Koomazaz** [dahall/TaskScheduler#40]
I am trying to create a task that runs under the "SYSTEM" user account, and I am able to do so using the RegisterTaskDefinition as such: ts.RootFolder.RegisterTaskDefinition(taskName, td, TaskCreation.CreateOrUpdate, "SYSTEM", null, TaskLogonType.ServiceAccount).  If I manually open up Windows Task Manager, I can see that it is still set to run as "SYSTEM."

However, if I create a TaskEditDialog and call ShowDialog for that task, the user property resets to the logged on user, rather than "SYSTEM."  I can see my user name in the "When running the task, use the following user account" field, and when I click OK, it updates the Task to use my user name.  However, if I click "Cancel," the Task (as visible in Windows Task Scheduler) still reflects the original user of "SYSTEM."

It seems as though the TaskEditDialog is intentionally resetting the user property on the Task to the currently-logged in user, disregarding the user that may have been purposefully set prior to the ShowDialog call.

**Comment by Koomazaz** [comment]
I think I found a work-around.  By setting the TaskDefinition.Principal.UserId property to "SYSTEM" also, the user does not get reset when calling TaskEditDialog.ShowDialog().

The original behavior still seems weird (maybe a bug), but this will allow us to work around the issue.

**Comment by dahall** [comment]
I'm in my debugger trying to recreate this problem and am not having success. Here's what I'm doing. Let me know if your steps are different:
1) Create a task as an administrator with the syntax you provided above. (RegisterTaskDef..."SYSTEM"...)
2) Open TaskEditDialog with task (`new TaskEditDialog(task, true, true)`) while logged on as admin
3) Look at user and see if changed from "SYSTEM"

On my system, running v2.7.2, it shows up as "SYSTEM", not the logged on user

**Comment by jotkol** [comment]
I had encounter similar account reset in my solution, too. 
Thare are two accounts named `user_name` - domain `DOMAIN_NAME\user_name` and local `user_name`.  In my solution, I add task prior to displaying edit dialog. Purpose is to not allow user to change the task's name because I use it later to identify task.
In the new task, before registering, I set TaskDefinition.Principal.UserId to `DOMAIN_NAME\user_name` and after registration it changes to just `user_name` and it's later recognized as local user name. Later call to `RegisterChanges()` or call to `RegisterTaskDefinition`  on dialog OK button (I tried both) fail because of **UnathorizedAccessException**. 

Simplified code:
```c#
    TaskService mgr = new TaskService();

    var def = mgr.NewTask();

    def.RegistrationInfo.Author = System.Security.Principal.WindowsIdentity.GetCurrent().Name;
    def.RegistrationInfo.Description = "Automatic task test";
    def.Principal.UserId = System.Security.Principal.WindowsIdentity.GetCurrent().Name; 

    def.Settings.Enabled = true;
    def.Settings.StartWhenAvailable = true;
    def.Settings.Hidden = false;

    var trigger = def.Triggers.AddNew(TaskTriggerType.Daily);
    trigger.StartBoundary = DateTime.Now;
    trigger.EndBoundary = DateTime.MaxValue;
    trigger.Enabled = true;

    var action = def.Actions.Add(path: "notepad.exe");

    // here newtask.Principal.UserId is "DOMAIN_NAME\user_name"

    var newtask = mgr.RootFolder.RegisterTaskDefinition(@"Test Daily Trigger " + Guid.NewGuid().ToString(), def, 
        TaskCreation.CreateOrUpdate, def.RegistrationInfo.Author, null, TaskLogonType.InteractiveToken);

    // here newtask.Principal.UserId is just "user_name"

    TaskEditDialog dlg = new TaskEditDialog(newtask, true, false); 
    var ret = dlg.ShowDialog(this);
    if (ret == DialogResult.OK)
        newtask.RegisterChanges(); // UnathorizedAccessException comes out of here
    else
        mgr.RootFolder.DeleteTask(newtask.Name);
```
During debugging session I noticed that TaskDefinition.Principal is recreated from IPrincipal data and this is the point ftom where domain name is no longer visible in Principal.

Another interesting facts are that when I add task programatically using domain account and go to Windows Task Scheduler and export it - task is assigned to domain user - it's ok. When I open it then save (no changes made at all) and export I see in XML file that user is now local. When I add task by hand in Windows Task Scheduler, similar user substitution during open-save occurs too.

**Comment by dahall** [comment]
@jotkol: Thank you for the code. I'm working through debugging now and will let you know my findings.

On how the XML file is behaving, I have found that a few things are different between the settings and the XML representations. Unfortunately, this library has no ability to affect this as it is part of the native Microsoft API.

**Comment by dahall** [comment]
@jotkol: While looking at your issue, there is something you should know that is outside of the problem you're seeing. The TaskEditDialog can take a TaskDefinition instance and a name and it will ensure that the task name is not changed. It is done as follows. Also, note that this way you don't have to capture the result since the dialog will register the task for you on OK.
```c#
var dlg = new TaskEditDialog(mgr, def, true, true, @"Test Daily Trigger" + Guid.NewGuid().ToString());
dlg.TaskNameIsEditable = false;
dlg.ShowDialog(this);
```
Also, all the properties you are setting for your trigger are the defaults (start, end and enabled). You can omit them all and just call `def.Triggers.Add(new DailyTrigger());`

When creating the task, along with setting the Principal.UserId property, you should also set the Principal.LogonType property to InteractiveToken. Or better, set it to S4U which is typically preferred due to UAC and privileges. This way you will get the exact UI you have without pre-registering the task.

Still working on duplicating your problem... More to come.

**Comment by dahall** [comment]
@jotkol: I think I found the problem. As you noticed, after registration, the domain name is removed. This is done by the native library and I cannot change it. However, in my TaskEditDialog, when receiving a username without a domain, I default to a local account if one is found. This is what created your problem. I have changed the logic so that it first checks for a domain account and will use that if found. There is a risk that I am artificially imposing a priority, but I think it is the majority need. All changes will be in release 2.7.4. By the way, I have reduced your code above to the following with the same outcome:
```c#
var def = TaskService.Instance.NewTask();
def.RegistrationInfo.Description = "Automatic task test";
def.Triggers.Add(new DailyTrigger());
def.Actions.Add("notepad.exe");
new TaskEditDialog(TaskService.Instance, def, true, true, @"Test Daily Trigger" +
   Guid.NewGuid().ToString()) { TaskNameIsEditable = false }.ShowDialog();
```

**Comment by dahall** [comment]
Without a response from @Koomazaz, and not being able to duplicate their issue with the current release, I am closing this issue.

**Comment by jotkol** [comment]
Thank you for your engagement. 

First. I use XML in order to read and compare different tasks properties in easy and unified way. Your note makes me not to trust this method. Sad.

Second. `dlg.TaskNameIsEditable = false;` is a cool feature that i've never found. I was too concentrated on some issues. :) `def.Triggers.Add(new DailyTrigger())` is a valuable hint, too.  I'm used to specify every property directly because of initial difficulties to add any working task to schedule. :)

Third. LogonType. Tasks, I want to define, should be used to make database backup. I want to specify task name, command, parameters, "run when user is not logged in" and a trigger initially. In main scenario a user  clicks OK and forgets about it. Sometimes he wants to change running account. I use InteractiveTokenOrPassword for easier testing but I think Password should be the best for that scenario.

Last. In spite of that domain name disappears after registration and xml correctness (XML again, sorry), int that moment inside definition XML element Task\Principals\Principal\UserId is set to domain user sid. This allow me to think that correct (or full) information about user account is stored somewhere. Why this cannot be obtain directly from Task Scheduler interfaces is MS secret, I think.
I'll try improvements soon.

**Comment by jotkol** [comment]
I don't know where to post now because of closed issue, but I get changes and I want to share.

I get current sources and I get exactly the same behaviour: **UnathorizedAccessException** on second registration because of wrong specified username. Only improvement I notice is better User class object creation in **TaskFolder.RegisterTaskDefinition** method and as result newly initialized _user_ variable has correct sid. Unfortunately this is never really used. Later call to native **RegisterTaskDefinition** utilizes `userId ?? user.Name` as parameter. This don't impact my case, becase userId equals 'user_name' and not null.
In this situation I've added UserId restoration line before opening the dialog

    newtask.Definition.Principal.UserId = System.Security.Principal.WindowsIdentity.GetCurrent().Name;

It does the trick. Now, always "fully qualified", user name is properly recognized during second registration.


Now I can refactor my code according to Your suggestions. :)

**Comment by jotkol** [comment]
TaskEditDialog constructor used to register new task (TaskDefinition) has disabled OK button. To enable it I did the trick based on source code reading. TaskName setter can work as OK-button-enabler if valid string provided:

    new TaskEditDialog(TaskService.Instance, def) {
        TaskNameIsEditable = false,
        TaskName = @"Test Daily Trigger" + Guid.NewGuid().ToString(),
    }.ShowDialog();

**Comment by dahall** [comment]
@jotkol I have re-opened the issue so we can keep track of it.

1. On your comment about using XML, I would still trust comparing XML to XML as that seems consistent. Where I see changes occurring is between the values in the native library and the XML file, like the user name.
2. I agree with your approach on using LogonType.
3. The domain name will get lost in the native library only. I have found that in the XML file *after* registration will contain the SID for domain accounts. That SID gets translated to just the user name in the library. I agree with your assessment that it must exist somewhere. I will look for it as part of this issue.
4. If I can find the problem behind the loss of the domain name, your workaround will not be needed.
5. [Updated] The OK button function is as designed. It will automatically enable once the "Name" field has a valid name. The reason your "trick" works is that it supplies a valid name!

**Comment by jotkol** [comment]
@dahall 

Ad. 5. You're right. The reason of this behaviour is I've tested back with public 2.7.3. With current source code setting task name in dialog constructor works OK. And no "trick" is needed. :)

**Comment by dahall** [comment]
@jotkol I think I have this fixed. Will you extract the contents of the zip files below and overwrite the assemblies in your project to test? This should always get the correct account for registered tasks, regardless of name conflicts.
* [TaskScheduler.zip](https://github.com/dahall/TaskScheduler/files/1786289/TaskScheduler.zip)
* [TaskSchedulerEditor.zip](https://github.com/dahall/TaskScheduler/files/1786290/TaskSchedulerEditor.zip)

Or, if you have a local NuGet repository, this zip file contains the 2 nupkg files
* [TaskScheduler.2.7.5.nupkg.zip](https://github.com/dahall/TaskScheduler/files/1786302/TaskScheduler.2.7.5.nupkg.zip)

**Comment by dahall** [comment]
As a note, the TaskPrincipal class (TaskDefintion.Principal) now has an Account property that will always get the fully qualified domain name by pulling it from the XML once the task is registered.

**Comment by jotkol** [comment]
@dahall Yeah. Now it works ok in discussed scenarios. 
I mean:
1. Registration prior to editing in dialog which register changed task too. This scenario is eventually just for test purposes only.
2. New task initializing and then editing and registration using dialog.

BTW, I want to add polish translation to the editor. How can I do that? I tried to add polish resources but there are a lot of files (46?) and string to change. I noticed you pushed russian translation in meantime so maybe there is a way to easy add such a base version for polish and then I could make an improvement. If not, I can make it myself, but it'll take a time :). My goal is to make it identical to system version.

**Comment by dahall** [comment]
@jotkol Thank you for validating the fixes.

On the Polish translation, I am running the initial translation now and will post a spreadsheet here for your validation in a few hours along with instructions. Thank you for your help in translation.

**Comment by jotkol** [comment]
@dahall No problem.
👍  I'm waiting...

**Comment by dahall** [comment]
Attached to this message is an Excel spreadsheet [TaskSched2018_03_08.xlsx](https://github.com/dahall/TaskScheduler/files/1793994/TaskSched2018_03_08.xlsx) with all the resource strings translated into the different languages. For the Polish translation in column ".pl", please add or correct each entry giving special attention to those strings in English that have format specifiers (e.g. {0}, {1} etc.) or that have an embedded '&' character to specify the keyboard shortcut or that use special characters ( or x->). The automated Azure translator usually messes those up. I would recommend using the Windows Task Scheduler tool as your guide for wording. The other localizations have copied it where possible for consistency. Microsoft has a [search engine](https://www.microsoft.com/en-us/language/Search) for localized phrases in Windows. If the cell is red, ignore it. Once done, post it here with your reply. I will then rebuild the project and release a new version with the new localization.

**Comment by dahall** [comment]
I have attached another version of the spreadsheet using a better translation engine. If you haven't already done much work, please use this one. 
[TaskSched2018_03_08.xlsx](https://github.com/dahall/TaskScheduler/files/1795348/TaskSched2018_03_08.xlsx)

**Comment by jotkol** [comment]
Ok. I take the second XLSX.

**Comment by dahall** [comment]
@jotkol Do you have an estimate as to when you will complete the Polish translations? If soon, I'll hold off pushing the next release of the library. If more than a few days, I'll go ahead and publish the next release with the automated translations and then will release again once I have your updates. No pressure either way, I'd just like to be responsive in getting updates published. Thank you.

**Comment by jotkol** [comment]
@dahall I am almost at the end now.

**Comment by jotkol** [comment]
@dahall I upload my work. [TaskSched2018_03_08_corrected_pl.xlsx](https://github.com/dahall/TaskScheduler/files/1803635/TaskSched2018_03_08_corrected_pl.xlsx)

**Comment by dahall** [comment]
All incorporated now. Both the account name fix and Polish translations with be included in 2.8.0 coming shortly.
